website/source/docs/auth/aws-iam.html.md

```diff
@@ -0,0 +1,606 @@
+---
+layout: "docs"
+page_title: "Auth Backend: AWS-IAM"
+sidebar_current: "docs-auth-aws-iam"
+description: |-
+  The aws-iam backend allows automated authentication AWS IAM principals.
+---
+
+# Auth Backend: aws-iam
+
+The aws-iam auth backend provides a secure introduction mechanism for AWS IAM
+principals. This allows you to reduce the problem of securely introducing a
+Vault token to the problem of securely introducing AWS IAM credentials, which
+AWS has already solved in a number of use cases, such as via EC2 Instance
+Profiles and IAM roles attached to Lambda functions. It also allows you to have
+a consistent workflow between developers working on a local laptop with tools
+such as Hologram.
+
+## Comparison with AWS-EC2 Authentication Backend
+
+The AWS-IAM and AWS-EC2 authentication backends serve similar purposes. Both
```

Comment:
> It seems to me that if we can determine whether a machine is an EC2 instance, the right thing to do here would be to merge this with the EC2 backend into simply an `aws` backend.
>
> That way we can support two types of login (instance identity doc or signed request). If the instance identity doc is used then we return a nonce and use the existing whitelisting capabilities and if this method is used then no nonce is needed (but may still be used/desired to allow only a single entity on an EC2 box to auth). Either way, if the identity proves to be an EC2 instance, we can apply the full set of binds.
>
> This would allow significant code reuse and allow this method to automatically gain support for further restrictions on EC2 instances.

Well, as of today, we can't determine definitively whether creds came from an EC2 instance in an IAM instance profile. We could guess it, or we could (better option, IMHO) give Vault administrators some sort of option to tell Vault it should make that inference. But we can't _prove_ it as of today (at least, not in a feasible way I know of).

More broadly, I don't think I'd want to merge the two authentication backends. I like the simplicity that each auth backend evaluates only a single method of asserting an identity (e.g., LDAP uses a username/password, GitHub uses a GH token, TLS uses x509 certs, etc.), as this simplicity makes it easier to reason about the security of a given Vault setup. Given some of the different security considerations between `aws-ec2` and the proposed `aws-iam` especially around replay attack prevention and nonces, I can easily see Vault administrators wanting to configure things like max TTLs differently between the two.

As an analogy, let's take the GitHub and LDAP auth backends. Let's further assume I have a GitHub Enterprise which happens to be using an LDAP backend for users. There's no reason in theory why I couldn't tell Vault to treat a user authenticated via the GitHub auth backend as an LDAP user and to do additional filtering based on LDAP attributes of those user. That doesn't mean the LDAP and GitHub backends should be merged -- they're very different ways of authenticating users with very different security properties (username/password vs. GH API token). But, to acheive code reuse, you should probably factor out the common functionality into shared utility functions to be used by the GitHub backend as well as the LDAP backend, and that would allow the GitHub backend to take advantage of new LDAP filters/restrictions as they are implemented.

In a similar vein, I view the `aws-ec2` backend as analogous to the LDAP backend, and the proposed `aws-iam` auth backend as analogous to the GitHub backend: you'd be trying to authenticate users using the GitHub or `aws-iam` native way and then infer that the authenticated entities correspond to objects in some other data store (an LDAP server or EC2 instances) and filter based off of that.

As I said earlier, Vault cannot tell on its own if the creds belong to an EC2 instance, though in theory an admin could tell Vault that it's safe to make that assumption. This also feels like more of a day 2 feature -- even without inferring an instance ID from the caller identity, this still solves important use cases as @jantman said. I'd like to get it out and get feedback before considering more sophisticated use cases (which I think this would be) and doing some refactoring. I've also reached out to our AWS contacts about the two things I mentioned (if `GetCallerIdentity` can return information on the actual session issuer and information on if it's MFA authenticated), and I think it makes sense to see if we can get a response back from them. (And if others following this conversation think these would be useful features and have AWS contacts, feel free to reach out as well and ask for these features.)